### ISSUE TITLE
Suggestion of better use of types to describe connection

### ISSUE THREAD

**Issue by amitaibu** [urish/elm-smart-bulb#1]
Hi @urish saw your tweet, so gave it a look :)

Usually an empty `String` is a bit of a code smell in Elm as types can be more expressive. Maybe there's a better way to describe that we still don't have the device name and ID -- using `Maybe`.

However I think you could model it even better by making impossible states impossible. That is, lets prevent having a Connected state without the name and ID.

So one could imagine something like this:

```elm
type alias Device = 
  {name : String, id : String}

type Connection = 
  NotConnected
  | Connected Device

Model = {connection: Connection}
```

However since you are dealing with [RemoteData](http://package.elm-lang.org/packages/krisajenkins/remotedata/latest) here, have just something like:

```elm
type alias Device = 
  {name : String, id : String}

Model = {connection: WebData Device}
```

So `NotAsked` == not connected, and `Success` will hold the device info.

**Comment by amitaibu** [comment]
I've edited above to be `Model = {connection: WebData Device}`

**Comment by urish** [comment]
Thank you so much @amitaibu, this is exactly the kind of feedback I was looking for :-)

I have adjusted the code - though I believe that the [update()](https://github.com/urish/elm-smart-bulb/blob/e8c78cdab8cf0b21a6856d5af43890f68dfce764/src/bulb.elm#L67) can still be improved. Is there any suggestion how to avoid the nested case there?

**Comment by amitaibu** [comment]
So here are a two more suggestions:

I think you should go with RemoteData . It's for two main reasons:

* It's a very common pattern, so it's good to know it
* As it has a `Failure err` state, you could wire those [errors](https://github.com/urish/elm-smart-bulb/blob/e8c78cdab8cf0b21a6856d5af43890f68dfce764/src/bluetooth.js#L35), and have a more robust solution. So your app will end up being more complete, but without too much effort.


While I've written `Model = {connection: WebData Device}` above I now see you get the device with another API, and not HTTP, so you could model your errors accordingly. Something like:

```elm
type BluetoothError = 
DeviceNotFound
| GeneralError String

type alias BluetoothData = RemoteData BluetoothError Device

Model = {connection: BluetoothData Device}
```
So for the errors, you'd probably have another port to send them to Elm, and you will decode it accordingly.

--

In general try always to Use types for readability and to help the compiler help you. For example in this [view](https://github.com/urish/elm-smart-bulb/blob/e8c78cdab8cf0b21a6856d5af43890f68dfce764/src/bulb.elm#L54-L57). You can have

```elm
type Color = Red | Green | Blue

type alias Rgb = (Int, Int, Int)

type State = 
  On Color
  | Off
```

Notice how the `Off` is not a color. It can later have a color, but it's a state.

So you can have a utility `getColorAndTextFromState : State -> (Rgb, String)` - that will return a tuple of the colors (e.g. `0xff, 0, 0`) and the text (e.g. `Red`)

**Comment by urish** [comment]
Thanks for the feedback! With some refactoring in place, I really like the new `Color`/`BulbState` types and the fact we now have some error handling (and can also represent the loading state). However, I am still unhappy with the `update()` function. Any ideas how to improve there?

**Comment by amitaibu** [comment]
Yeah this is definitely getting into a much nicer shape!

> However, I am still unhappy with the update() function. Any ideas how to improve there?

It's more common to see a single `case msg of` and then handling the values there. I guess you are somewhat unhappy because you can't restrict the `msg` to a specific state of the model. But indeed, it won't be easy. Maybe you could do some type trickery, but it would be akward .

I think it would be more readable to deal with the state per `msg`, so it would look something like:

```elm
update : Msg -> Model -> ( Model, Cmd Msg )
update msg model =
    let
        noChange =
            ( model, Cmd.none )
    in
    case msg of
        RequestDevice ->
            if isNotAsked model.connection then
                ( Model Loading
                , Bluetooth.requestDevice bulbService
                )
            else
                noChange
```

---

use [elm-format](https://github.com/avh4/elm-format#experimental-version) (the `@exp` is great and stable)

---

[elm-html-test](http://package.elm-lang.org/packages/eeue56/elm-html-test/latest) is lovely for unit testing. I usually use it on the `view` function itself. so I can test a lot of my logic.

Here's one that constantly blows my mind. Those [docs](https://github.com/Gizra/elm-editable-webdata/blob/52e02c4a9d24232cb84fa019a0991f0fc2ceb013/src/Editable/WebData.elm#L55-L68) I've added, are [converted](https://github.com/Gizra/elm-editable-webdata/blob/52e02c4a9d24232cb84fa019a0991f0fc2ceb013/execute-tests#L3) into unit tests. So the docs are always right!

---

Nitpick: bulb.elm -> Bulb.elm. Also, checkout the [structure](https://github.com/Gizra/drupal-elm-starter/tree/master/client) we start our apps. It's divided to the `Model/ Update/View/Utils/Test`

**Comment by exp** [comment]
> the @exp is great and stable

How kind of you :)

**Comment by amitaibu** [comment]
Yeah, sorry @exp I didn't know if I should apologise, or let that accidental mentioning slip by 😉 

Anyway, Elmlang is super recommended 😸

**Comment by exp** [comment]
Don't worry about it, it happens *a lot*!

I like everything about Elm other than its compilation target :smirk_cat:

**Comment by amitaibu** [comment]
> Don't worry about it, it happens a lot!

I can imagine!

Well, one day it's going to target wasm, so stay tuned ..

**Comment by urish** [comment]
thanks @amitaibu for all the suggestions! I started using `elm-format`, and will check out the recommended structure you linked.

I also refactored update and [tried my luck](https://github.com/urish/elm-smart-bulb/blob/31f720f75a8b2e9dead6a1dd2a354d84043ec4fa/src/Bulb.elm#L188) with `Maybe` - it does look better to me. What do you think?

Eventually, I hope to grow [Bluetooth.elm](https://github.com/urish/elm-smart-bulb/blob/31f720f75a8b2e9dead6a1dd2a354d84043ec4fa/src/Bluetooth.elm) into a generic Web Bluetooth Elm library, so any guidance on how to get started with this will be much appreciated

**Comment by urish** [comment]
and hi @exp :-) 
how many people have you met so far by accidentally getting tagged?

**Comment by amitaibu** [comment]
App looks great now 👍 

(You can probably remove `getDevice`, in favor of the more generic `RemoteData.toMaybe`)

I encourage you to add a unit test, and hook it into travis. I find that projects with the Travis badge always look more legit than the ones without 😉 

Welcome to Elm!

**Comment by urish** [comment]
I love `RemoteData.toMaybe`, just applied it, and also set up testing + travis :)

next up - I will check elm-html-test that you mentioned. Thanks for all the guidance so far!
